cpp/src/precompiled/hash.cc

```diff
@@ -0,0 +1,280 @@
+// Copyright (C) 2017-2018 Dremio Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern "C" {
+
+#include <string.h>
+#include "./types.h"
+
+static inline uint64 rotate_left(uint64 val, int distance) {
+  return (val << distance) | (val >> (64 - distance));
+}
+
+//
+// MurmurHash3 was written by Austin Appleby, and is placed in the public
+// domain.
+// See http://smhasher.googlecode.com/svn/trunk/MurmurHash3.cpp
+// MurmurHash3_x64_128
+//
+static inline uint64 fmix64(uint64 k) {
+  k ^= k >> 33;
+  k *= 0xff51afd7ed558ccduLL;
+  k ^= k >> 33;
+  k *= 0xc4ceb9fe1a85ec53uLL;
+  k ^= k >> 33;
+  return k;
+}
+
+static inline uint64 murmur3_64(uint64 val, int32 seed) {
+  uint64 h1 = seed;
+  uint64 h2 = seed;
+
+  uint64 c1 = 0x87c37b91114253d5ull;
+  uint64 c2 = 0x4cf5ad432745937full;
+
+  int length = 8;
+  uint64 k1 = 0;
+
+  k1 = val;
+  k1 *= c1;
+  k1 = rotate_left(k1, 31);
+  k1 *= c2;
+  h1 ^= k1;
+
+  h1 ^= length;
+  h2 ^= length;
+
+  h1 += h2;
+  h2 += h1;
+
+  h1 = fmix64(h1);
+  h2 = fmix64(h2);
+
+  h1 += h2;
+
+  // h2 += h1;
+  // murmur3_128 should return 128 bit (h1,h2), now we return only 64bits,
+  return h1;
+}
+
+static inline uint64 double_to_long_bits(double value) {
+  uint64 result;
+  memcpy(&result, &value, sizeof(result));
+  return result;
+}
+
+FORCE_INLINE int64 hash64(double val, int64 seed) {
+  return (int64)murmur3_64(double_to_long_bits(val), (int32)seed);
+}
+
+FORCE_INLINE int32 hash32(double val, int32 seed) {
+  return (int32)murmur3_64(double_to_long_bits(val), seed);
+}
+
+// Wrappers for all the numeric/data/time arrow types
+
+#define HASH64_WITH_SEED_OP(NAME, TYPE)                                              \
+  FORCE_INLINE                                                                       \
+  int64 NAME##_##TYPE(TYPE in, boolean is_valid, int64 seed, boolean seed_isvalid) { \
```

Comment:
Shouldn't there be a check for seed_isvalid?

Reply:
the seed is actually a literal. but, our function type (NULL_NEVER/NULL_INTERNAL) require the is_valid bit to be passed for every param. 

checking for seed_isvalid is harmless, so I'll add it.